File: src/main/java/org/academy/api/client/gui/render/Position2D.java
package org.academy.api.client.gui.framework.render;

import org.joml.Vector2f;

import static org.academy.api.common.util.MathUtil.Axis2D;
import static org.academy.api.common.util.MathUtil.Direction2D;

public final class Position2D extends Vector2f {
    public Position2D(float x, float y) {
        super(x, y);
    }

    public Position2D() {
    }

    public static Position2D of(Axis2D axis, float primaryPosition, float secondaryPosition) {
        return switch (axis) {
            case HORIZONTAL -> new Position2D(primaryPosition, secondaryPosition);
            case VERTICAL -> new Position2D(secondaryPosition, primaryPosition);
        };
    }

    public Position2D step(Direction2D direction) {
        return switch (direction) {
            case DOWN -> new Position2D(x, y + 1);
            case UP -> new Position2D(x, y - 1);
            case LEFT -> new Position2D(x - 1, y);
            case RIGHT -> new Position2D(x + 1, y);
        };
    }

    public float getCoordinate(Axis2D axis) {
        return switch (axis) {
            case HORIZONTAL -> x;
            case VERTICAL -> y;
        };
    }
}

File: src/main/kotlin/org/academy/api/client/gui/render/Position2D.kt
package org.academy.api.client.gui.render

import org.academy.api.common.util.MathUtil.Axis2D
import org.academy.api.common.util.MathUtil.Direction2D
import org.joml.Vector2f

class Position2D : Vector2f {
    constructor(x: Float, y: Float) : super(x, y)

    constructor()

    fun step(direction: Direction2D): Position2D {
        return when (direction) {
            Direction2D.DOWN -> Position2D(x, y + 1)
            Direction2D.UP -> Position2D(x, y - 1)
            Direction2D.LEFT -> Position2D(x - 1, y)
            Direction2D.RIGHT -> Position2D(x + 1, y)
        }
    }

    fun getCoordinate(axis: Axis2D): Float {
        return when (axis) {
            Axis2D.HORIZONTAL -> x
            Axis2D.VERTICAL -> y
        }
    }

    companion object {
        fun of(axis: Axis2D, primaryPosition: Float, secondaryPosition: Float): Position2D {
            return when (axis) {
                Axis2D.HORIZONTAL -> Position2D(primaryPosition, secondaryPosition)
                Axis2D.VERTICAL -> Position2D(secondaryPosition, primaryPosition)
            }
        }
    }
}